openlr/openlr_match_quality/openlr_assessment_tool/mainwindow.cpp

```diff
@@ -38,50 +49,184 @@ class TrafficDrawerDelegate : public TrafficDrawerDelegateBase
+                       .Width(4.0f).ShowPoints(true /* markPoints */));
+  }
+
+  void ClearGoldenPath() override
+  {
+    m_drapeApi.RemoveLine(kGoldenLineId);
+  }
+
+  void ClearAllPaths() override
   {
     m_drapeApi.Clear();
   }
 
-private:
-  string NextLineId() { return strings::to_string(m_lineId++); }
 
-  uint32_t m_lineId = 0;
+  void VisualizePoints(std::vector<m2::PointD> const & points) override
+  {
+    UserMarkControllerGuard g(m_bm, UserMarkType::DEBUG_MARK);
+    g.m_controller.SetIsVisible(true);
+    g.m_controller.SetIsDrawable(true);
+    for (auto const & p : points)
+      g.m_controller.CreateUserMark(p);
+  }
+
+  void ClearAllVisualizedPoints() override
+  {
+    UserMarkControllerGuard g(m_bm, UserMarkType::DEBUG_MARK);
+    g.m_controller.Clear();
+  }
 
+private:
   Framework & m_framework;
   df::DrapeApi & m_drapeApi;
+  BookmarkManager & m_bm;
+};
+
+bool PointsMatch(m2::PointD const & a, m2::PointD const & b)
+{
+  auto constexpr kToleraneDistanceM = 1.0;
+  return MercatorBounds::DistanceOnEarth(a, b) < kToleraneDistanceM;
+}
+
+class PointsControllerDelegate : public PointsControllerDelegateBase
+{
+public:
+  PointsControllerDelegate(Framework & framework)
+    : m_framework(framework)
+    , m_index(framework.GetIndex())
+    , m_roadGraph(m_index, routing::IRoadGraph::Mode::ObeyOnewayTag,
+                  make_unique<routing::CarModelFactory>())
+  {
+  }
+
+  std::vector<m2::PointD> GetAllJunctionPointsInViewPort() const
+  {
+    std::vector<m2::PointD> points;
+    auto const & rect = m_framework.GetCurrentViewport();
+    auto const pushPoint = [&points, &rect](m2::PointD const & point) {
+      if (!rect.IsPointInside(point))
+        return;
+      for (auto const & p : points)
+      {
+        if (PointsMatch(point, p))
+          return;
+      }
+      points.push_back(point);
+    };
+
+    auto const pushFeaturePoints = [&pushPoint](FeatureType & ft) {
+      if (ft.GetFeatureType() != feature::GEOM_LINE)
+        return;
+      auto const roadClass = ftypes::GetHighwayClass(ft);
+      if (roadClass == ftypes::HighwayClass::Error ||
+          roadClass == ftypes::HighwayClass::Pedestrian)
+      {
+        return;
+      }
+      ft.ForEachPoint(pushPoint, scales::GetUpperScale());
+    };
+
+    m_index.ForEachInRect(pushFeaturePoints, rect, scales::GetUpperScale());
+    return points;
+  }
+
+  std::vector<FeaturePoint> GetFeaturesPointsByPoint(m2::PointD & p) const
+  {
+    std::vector<FeaturePoint> points;
+    m2::PointD pointOnFt;
+    indexer::ForEachFeatureAtPoint(m_index, [&points, &p, &pointOnFt](FeatureType & ft) {
+        if (ft.GetFeatureType() != feature::GEOM_LINE)
+          return;
+        size_t pointIndex = 0;
+        auto minDistance = std::numeric_limits<double>::max();
+
+        ft.ForEachPoint(
+            [&points, &p, &ft, &pointIndex, &minDistance, &pointOnFt](m2::PointD const & fp)
```

Comment:
Please, replace the capture list by a single ```&```. So precise specification is not needed here - the callback will be called "right there, right now", but so detailed capture list complicates understanding.